[FLINK-15911][runtime] Make Flink work with NAT.

xintongsong opened the pull request:
## What is the purpose of the change

This PR makes Flink work with NAT by introducing separated configuration options for address/port and bind-address/bind-port of JM/TM RPC services and Netty Shuffle Service.

## Brief change log

- 5cb9b8065bbf6bf6f9445cb20b4d7e984688e468: Minor code clean-ups.
- 1fcee59ad6fb19a8e6f73b648fb635c939612e5f: Deduplicate codes in docker e2e tests.
- b3e88d6c08f91698136d6f150750592955a15f2e: Refactor to create `AkkaRpsService` with builder class. Currently there are too many nested methods for creating a `AkkaRpsService`, making it hard to understand which changes affect which code paths.
- 3d5e3f6cf4b800a9e43c88552a531122bbae3ad5: Separate config options for JM/TM hostname/bind-hostname and RPC port/bind-port.
- d2e5f15a8823a2de0aec24764a9cc93b75f0deaa: Separate config options for Netty Shuffle Service data port/bind-port.
- a8d628c0a325485aa33c7beed94b1eea9d880faa: Introduce e2e test for Flink over NAT.
- 6e0a577481f9300c87299b0629fab6dc1b3bd71a: **DO_NOT_MERGE_THIS** Add the NAT e2e test to pre-commit tests, for the convenience of review.

## Verifying this change

Added a docker-based e2e test for verifying Flink works over NAT.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not documented)

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/rpc/akka/AkkaRpcServiceBuilder.java:
```diff
@@ -0,0 +1,136 @@
+/**
+ * Builder for {@link AkkaRpcService}.
+ */
+public class AkkaRpcServiceBuilder {
```
I don't get it.
- Constructor of `AkkaRpcService` is used in various of test cases. Unless we want to modify them all, I don't think we can make the constructor private. Maybe it makes sense to annotate the constructor as `VisibleForTesting`, because without the test cases it can be package visibility.
- If we cannot make constructor of `AkkaRpcService` private anyway, then I would prefer to put `AkkaRpcServiceBuilder` in a separate file. The file `AkkaRpcService.java` is already quite long. No need to further complicate it.
- I don't see how `AkkaRpcService#local(remote)RpcServiceBuilder` is more readable compared to `AkkaRpcServiceUtils#local(remote)ServiceBuilder`.
- It probably make sense to make the constructor of the builder private, since we already have `AkkaRpcServiceUtils#local(remote)ServiceBuilder`. In that case, I would tend to put the builder inside `AkkaRpcServiceUtils` rather than `AkkaRpcService`.

KarmaGYZ replied:
I think whether we make it an inner class or not should only depend on whether we need to leverage the characteristic of the inner class mechanism. It has nothing to do with how many tests we need to modify or how complicated the `AkkaRpcService` is.

In this case, we could get benefit from making it an inner class of `AkkaRpcService` is:
- For both AkkaRpcService and Builder, we could prevent unexpected access to their constructor.
- We could expose two static functions whose use case is well explained by the method name, instead of two constructors only different in param list.

The alternative, which annotates the constructor as `VisibleForTesting` and put the builder inside `AkkaRpcServiceUtils`, also LGTM. In this way, we could only build an `AkkaRpcService` through the utility class. If it needs lots of effort to change the test cases, I think it would be good.

xintongsong replied:
I think whether we want to make a constructor private, depends on whether allowing it to be accessed from outside may cause problem. E.g., the none argument constructor of `ResourceProfile` must be private because the only case we need it is for constructing the singleton instance `UNKNOWN`. If we allow to create another instance of `ResourceProfile` with this none argument constructor, the created instance may not work properly.

In this case, I don't see such problem in allowing direct access to the constructors. That makes hiding the constructors a nice but not must to have approach, by which we suggest people to choose the more convenient builders. 

On the other hand, it is not about how many efforts needed that we do not use builders in the test cases. Currently, the builder does two things: create a `ActorSystem`, and use it to create a `AkkaRpcService`. However, in many test cases, we need to use a custom testing actor system, which is different from the one we created for production. That's way we cannot easily use builders in the tests.

We could also make the builder to support custom actor system. But that complicates the builder logics (what should we do if both custom actor system and the fields for creating the production actor system are provided), and I think it doesn't make sense to do for pure testing purpose.

xintongsong replied:
I moved `AkkaRpcServiceBuilder` into `AkkaRpcServiceUtils`, and made its constructors private. Also annotated constructor of `AkkaRpcService` `VisibleForTesting`.

xintongsong commented on flink-end-to-end-tests/test-scripts/test_nat.sh:
```diff
@@ -0,0 +1,71 @@
+
+# user inside the container must be able to create files, this is a workaround in-container permissions
+mkdir -p $OUTPUT_VOLUME
+chmod 777 $OUTPUT_VOLUME
```
I think here we simply give the container full access to the path, not assuming what it does with the path. What is the problem?

KarmaGYZ replied:
It does not infect the correctness of this test case. My point is we need to be careful when setting the permission of the directory mount in the container, in case anything happens unexpectedly while mutely.

xintongsong replied:
I would argue that this is a new created empty directory dedicated for the test case to use, and will be be cleaned up when the test is finished. But never mind, I'll make it `666` if it bothers you.

xintongsong replied:
Turns out `666` does not work. Flink throws the following error:

> Output directory '/data/test/output/docker_wc_out' could not be created.

And I found from google that:

> For directories, execute permission allows you to enter the directory (i.e., cd into it), and to access any of its files. http://web.mit.edu/sipb/doc/working/afs/html/subsection3.1.html

KarmaGYZ replied:
My bad. Let's keep it to 777.

xintongsong commented on flink-end-to-end-tests/test-scripts/test_nat.sh:
```diff
@@ -0,0 +1,71 @@
+docker-compose -f ${DOCKER_SCRIPTS}/docker-compose.nat.yml logs taskmanager1 > ${FLINK_DIR}/log/taskmanager1.log
+docker-compose -f ${DOCKER_SCRIPTS}/docker-compose.nat.yml logs taskmanager2 > ${FLINK_DIR}/log/taskmanager2.log
+
+cat $OUTPUT_VOLUME/docker_wc_out/1 $OUTPUT_VOLUME/docker_wc_out/2 | sort > $OUTPUT_VOLUME/docker_wc_out/merged
```
`check_result_hash_no_exit` requires a common prefix of the output files, which the outputs of WordCount do not have.

KarmaGYZ replied:
The prefix could be "$OUTPUT_VOLUME/docker_wc_out/" here. BTW, we could extract `docker_wc_out` as a parameter called `OUTPUT_DIR` or `OUTPUT_FILE`.

xintongsong replied:
You're right.
I think I tried this before, but forgot to add the '/' in the prefix so it failed.

xintongsong commented:
Thanks @KarmaGYZ, I've addressed the comments, and travis has gave a green light. https://travis-ci.org/xintongsong/flink/builds/657623133

@tillrohrmann, @zhijiangW Could you please also help take a look at this PR? Thanks

xintongsong commented:
@flinkbot run azure

xintongsong commented:
@tillrohrmann 
I checked the test failures.

The failures on Azure are all unrelated.

The NAT e2e test failure on Travis is cause by error message in logs.
- The job is executed successfully, with correct result outputted.
- The error message is about the RM RPC service not started yet when one of the TMs tries to connect to it. 
- The following JM log shows that the message is successfully received before discarded, indicating the TM has no problem resolving the correct RM address and RPC port.
`The rpc endpoint org.apache.flink.runtime.resourcemanager.StandaloneResourceManager has not been started yet. Discarding message org.apache.flink.runtime.rpc.messages.RemoteFencedMessage until processing is started.`
- I think this is not a real problem, because TM will retry connecting to RM later.

I've set `skip_check_exceptions` for this test case. I think relying on the result hash check should be enough.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/BootstrapTools.java:
```diff
@@ -82,92 +82,80 @@
+			String externalAddress,
+			String externalPortRange,
+			String bindAddress,
+			int bindPort,
```
Actually, I also thought about making the bindPort Optional, and maybe also bindAddress.

However I also noticed that the code style guide suggests to use Optional only for return values of API/public methods, but not for class fields and method arguments.

Not sure if it's ok to not strictly following the guide. We do have a good reason to use Optional here.

tillrohrmann replied:
I think we said that it is ok to use `Optional` as a parameter for internal functions if it makes things easier/clearer. Hence, it should be ok to use it here.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java:
```diff
@@ -261,26 +258,30 @@ public void start() throws Exception {
 					// start a new service per component, possibly with custom bind addresses
 					final String jobManagerBindAddress = miniClusterConfiguration.getJobManagerBindAddress();
 					final String taskManagerBindAddress = miniClusterConfiguration.getTaskManagerBindAddress();
+					final String jobManagerBindPort = miniClusterConfiguration.getJobManagerBindPortRange();
```
No, and I don't think we should.

Whenever the user needs to specify a bind port apart form the external port, the user also needs to set port forwarding rules in the environment, e.g. NAT gateway. For each external port, it can only be forwarded to one internal port. Therefore, it does not make sense to support a range of bind ports, because the user needs to know which internal port is bound anyway.

tillrohrmann replied:
Ok, then we should maybe change `getJobManagerBindPortRange` into `getJobManagerBindPort` to better reflect this.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java:
```diff
@@ -261,26 +258,30 @@ public void start() throws Exception {
 					final String jobManagerBindAddress = miniClusterConfiguration.getJobManagerBindAddress();
 					final String taskManagerBindAddress = miniClusterConfiguration.getTaskManagerBindAddress();
+					final String jobManagerBindPort = miniClusterConfiguration.getJobManagerBindPortRange();
+					final String taskManagerBindPort = miniClusterConfiguration.getTaskManagerBindPortRange();
```
Please see my explanation above.

tillrohrmann replied:
Same here: Please rename `getTaskManagerBindPortRange` into `getTaskManagerBindPort`.

xintongsong commented on flink-core/src/main/java/org/apache/flink/configuration/JobManagerOptions.java:
```diff
@@ -54,6 +54,16 @@
+	 * The local address of the network interface that the job manager binds to.
+	 */
+	public static final ConfigOption<String> BIND_HOST =
+		key("jobmanager.bind-host")
```
This address is not only used by the RPC service, but also shared by the blob server, rest server, and potentially any other service in future that needs to bind to the network interface.

To that end, I think the config option for TM ("taskmanager.host") makes more sense. I guess that might be the reason we use "host" in the config key at the first place.

tillrohrmann replied:
makes sense. Thanks for the clarification.

xintongsong commented on flink-core/src/main/java/org/apache/flink/configuration/TaskManagerOptions.java:
```diff
@@ -94,19 +94,28 @@
+	 * The local address of the network interface that the task manager binds to.
+	 */
+	public static final ConfigOption<String> BIND_HOST =
+		key("taskmanager.bind-host")
```
Same here. TM address and bind-address are shared by RPC service, netty shuffle service and queryable kv state.

I'm actually thinking about changing "jobmanager.rpc.address" to "jobmanager.host" as a follow up, to make the config keys consistently using "host".

tillrohrmann replied:
ok this makes sense.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/ClusterEntrypoint.java:
```diff
@@ -248,10 +247,14 @@ protected void initializeServices(Configuration configuration) throws Exception
+				configuration.getString(JobManagerOptions.BIND_HOST),
+				configuration.contains(JobManagerOptions.RPC_BIND_PORT) ?
+					configuration.getInteger(JobManagerOptions.RPC_BIND_PORT) :
+					null);
```
Same here. I also think using Optional will express the contract better. The question is do we need to strictly follow the code style guide.

tillrohrmann replied:
Since it is an internal method, we can diverge a bit from the code style here.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniClusterConfiguration.java:
```diff
@@ -102,21 +102,21 @@ public int getNumTaskManagers() {
-	public String getTaskManagerBindPortRange() {
-		return configuration.getString(TaskManagerOptions.RPC_PORT);
+	public int getTaskManagerBindPort() {
+		return configuration.getInteger(TaskManagerOptions.RPC_BIND_PORT, 0);
```
Not sure if I fully understand your concern.

I think the RPC services are meant to be used by Flink components internally. Then how would user be affected if the configured address is not used internally?

I have no problem preserving the external address/ports if configured. Just trying to understand.

tillrohrmann replied:
I think the problem I see is that we are using `RPC_BIND_PORT` and `BIND_HOST` as the external address and port when creating the `RpcService` in the `MiniCluster`. It might not be a big problem but it is a behaviour inconsistency.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskManagerRunner.java:
```diff
@@ -351,21 +351,21 @@ public static TaskExecutor startTaskManager(
 		LOG.info("Starting TaskManager with ResourceID: {}", resourceID);
 
-		InetAddress remoteAddress = InetAddress.getByName(rpcService.getAddress());
+		InetAddress externalAddress = InetAddress.getByName(rpcService.getAddress());
```
True. I'll try to fix the problem and update the e2e test case.

xintongsong replied:
After looking more into this, I'm not entirely sure whether we should support unresolvable external addresses.

The prerequisites are unclear to me. If an external address cannot be resolved by the TaskManager itself, then where can it be resolved? Do we assume all other components (JobManager and other TaskManagers) except for itself can resolve the external address? That sounds like a strong assumption to me, which may not always be true. E.g., deploying a Flink cluster over machines from two different sub-network, each sub-network has multiple machines/TMs. TMs in the same sub-network might not be able to resolve each other's external address. Another example, multiple TMs on the same machine.

I don't see a good way to cover all such cases. If we can only support the special case where external addresses can be resolved by all other components, I'm not sure whether we should do it given the efforts needed. Alternatively, user might use IP addresses to workaround in such cases.

tillrohrmann replied:
I thought that this is actually the problem we are trying to solve with this PR. Assume the `TM` is running in a Docker container which has an external name which is not resolvable from within the container. If all other Flink components can resolve this name, then it would work if the `TM` itself would not try to resolve the address.

At the moment we have the requirement that all `TM` components can resolve their own and the addresses of the other `TMs`. I think this requirement is too strict. What we actually need is that for a given `TM` `t` all other `TMs` `t'` can resolve the address of `t`. 

Of course, this means that if the user deploys multiple `TMs` into the same network where they cannot resolve each others external names, then Flink would not work. But this is ok because it violates the requirement that other `TMs` need to be able to resolve a `TM's` address. It is then the user's responsibility to configure the cluster accordingly.

Moreover, one could still enforce the current model by configuring the `bind-host` to the external address. If this is the case, then a `TM` would only start if it can resolve its own address.

xintongsong replied:
Alright, thanks for the explanation. I'll try to fix it.

wangyang0918 replied:
@tillrohrmann @xintongsong 
FYI
I think we do not have the requirement that other `TMs` need to be able to resolve a `TM's ` address. For current standalone Flink cluster on K8s, all the `TMs` could not resolve other's hostname. However, it works pretty well. This is the case for current master code without external address.

When  introducing the external address, i think it is same that we do not require other `TMs` need to be able to resolve a `TM's ` address.

Correct me if i am wrong somewhere.

tillrohrmann replied:
I think it currently works because the TM will resolve its own hostname and communicate the resolved IP address to the other TMs. If the TM cannot resolve it, then it will fail. If we introduce an external hostname, then this will no longer be possible.

wangyang0918 replied:
I agree with you that if the `TM` could not resolve its own address, it will fail. Indeed, it is problem we need to fix in this PR. However, i am just wondering given that we have fix this and do not require the `TM` could resolve its own external hostname, do we still need other `TMs` could resolve the external hostname? 

In my opinion every `TM` registers to the `JM` with hostname and ip. So all the `TMs` could get other `TM` hostname and ip. I am not sure about this.

tillrohrmann replied:
The problem with the external address is that it might resolve to an IP which is not known to the `TM`. This, of course, could also point towards a broken setup but if people want to run a `TM` behind a NAT, this could be a problem.

xintongsong commented:
Thanks for the review @tillrohrmann.

> A quick question for my understanding. In the e2e test we give every container an external address which is equal to the hosts ip address, right? How can the docker container route the packages for this address if the docker network does not happen to use the same subnet as the host's IP address? Can docker container talk to services running in the host's network? I guess I don't understand Docker good enough...

As I understand, all the packages whose destination is not the docker network will be routed to the host. It depends on the host's setting whether and where to further route the package, which, without any special settings, should be simply forward the package like any other package sent from the host. Accessing other services running on the host should have no difference from accessing any public Internet address from inside a docker container.

tillrohrmann commented:
Thanks for the clarification. It makes sense to me now :-)

Concerning the external address resolution on the `TaskManager`, I've responded to the respective comment. I think we need to clarify what we want to achieve with this PR.

xintongsong commented:
Thanks for the responses, @tillrohrmann.
I'll try to address the comments asap, including the unresolvable hostname issue.
BTW, I think you overlooked two comments regarding the binding port range, which are in the "hidden conversations".

tillrohrmann commented:
True, I've responded two the other comments. I would suggest a simple renaming of the methods.

xintongsong commented:
@tillrohrmann, I have addressed your comments.

However, I'm still trying to understand how the following is a common need.

> What we actually need is that for a given TM `t` all other TMs `t'` can resolve the address of `t`.

I find it not natural to simulate the above scenario with Docker in the e2e test. For each container I have to manually add the hostname-ip entries of all other containers except for itself.

Say all the JM and TMs are running in Docker containers. If a given TM `t` cannot resolve its own external address from inside the container, then unless specifically configured like I did in the e2e test, how could the JM and all the other TMs `t'`, which are also inside containers, resolve the address of `t`? If `t'` can resolve the address of `t` by using the same DNS settings as its host, then shouldn't `t'` also use the same DNS settings as its host and be able to resolve its own address?

xintongsong commented:
@flinkbot run travis re-run the last Travis build
@flinkbot run azure re-run the last Azure build

tillrohrmann commented:
> @tillrohrmann, I have addressed your comments.
> 
> However, I'm still trying to understand how the following is a common need.
> 
> > What we actually need is that for a given TM `t` all other TMs `t'` can resolve the address of `t`.
> 
> I find it not natural to simulate the above scenario with Docker in the e2e test. For each container I have to manually add the hostname-ip entries of all other containers except for itself.
> 
> Say all the JM and TMs are running in Docker containers. If a given TM `t` cannot resolve its own external address from inside the container, then unless specifically configured like I did in the e2e test, how could the JM and all the other TMs `t'`, which are also inside containers, resolve the address of `t`? If `t'` can resolve the address of `t` by using the same DNS settings as its host, then shouldn't `t'` also use the same DNS settings as its host and be able to resolve its own address?

I would assume that in most setups every `t` would be able to resolve all of the others `t'` addresses and also its own. But this is a stronger assumption than saying every `t'` needs to be able to resolve 't' address and hence it breaks some setups (as people complained on the ML/SO).

xintongsong commented:
@tillrohrmann

I'm ok with supporting the scenarios where address of `t` can be resolved by all other `t'` except for itself, and have updated the PR to address this issue.After all, it is indeed an improvement that we are able to support more use cases without introducing any problem to exist ones.

It just how tricky I have to simulate this scenario in the e2e test that makes me suspect something might be wrong. Or put it another way, why users want to use Flink like this? Have we truly understood what they need?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/KvStateService.java:
```diff
@@ -170,7 +170,7 @@ public static KvStateService fromConfiguration(TaskManagerServicesConfiguration
 				taskManagerServicesConfiguration.getNumberOfSlots() : qsConfig.numProxyQueryThreads();
 			kvClientProxy = QueryableStateUtils.createKvStateClientProxy(
-				taskManagerServicesConfiguration.getTaskManagerAddress(),
+				taskManagerServicesConfiguration.getExternalAddress(),
```
Not sure, whether this change is valid. The problem is that the underlying Netty server uses this address to bind against. If the external address is not resolvable, it might fail.

xintongsong replied:
As mentioned in the JIRA ticket, I'd like to exclude queryable state from the scope of this PR. Because I'm not familiar with how it works, and it seems does not have many use cases.

I think this should not affect other use cases, because the code path should only be executed if queryable state is enabled.

I think we should mention in the release notes that the new NAT support does not cover queryable state.

tillrohrmann replied:
Sounds good.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java:
```diff
@@ -261,26 +258,30 @@ public void start() throws Exception {
+					final String taskManagerBindPort = miniClusterConfiguration.getTaskManagerBindPortRange();
+
+					dispatcherResourceManagreComponentRpcServiceFactory =
+						new DedicatedRpcServiceFactory(configuration, jobManagerBindAddress, jobManagerBindPort);
```
I think you are right. A `DedicatedRpcServiceFactory` might be used to create multiple RPC services. In such cases, the current implementation will lead to conflicts on binding ports. To that end, I would agree with not supporting the binding port option and rely on the random port.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/KvStateService.java:
```diff
@@ -181,7 +181,7 @@ public static KvStateService fromConfiguration(TaskManagerServicesConfiguration
 			int numStateServerQueryThreads = qsConfig.numStateQueryThreads() == 0 ?
 				taskManagerServicesConfiguration.getNumberOfSlots() : qsConfig.numStateQueryThreads();
 			kvStateServer = QueryableStateUtils.createKvStateServer(
```
I have no opinion on this.

I'm not familiar with how the queryable state works and its use cases. I've consulted @zhijiangW about this, and he suggested that queryable state does not have much use cases and may not be necessary for the NAT scenarios.

In addition, if we want to support this for queryable state, we could use some help from people who are more familiar with the queryable state.

tillrohrmann replied:
Sounds good.

xintongsong commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMaster.java:
```diff
@@ -571,9 +572,16 @@ private void releaseEmptyTaskManager(ResourceID resourceId) {
 
+		final TaskManagerLocation taskManagerLocation;
+		try {
+			taskManagerLocation = TaskManagerLocation.fromUnresolvedLocation(taskManagerUnresolvedLocation);
```
For data transmission, yes. But I think the TM's address is required to be resolvable for JM process anyway. Because we still need them for RPC communications.

tillrohrmann replied:
I guess it should be fine since the `TM` establishes the connection. But this is hairsplitting from my side. I'm good with keeping it as is.

xintongsong commented:
@tillrohrmann Thanks for the review. I've addressed the comments. 

Regarding queryable state, I did take a look but couldn't find an easy way to separate the external and bind address/port for it. It uses Netty `ServerBootstrap` (in `AbstractServerBase`), which seems not providing any API for separating external & bind. I guess it requires more background knowledge than I have to touch this feature.
